content/specs/xrpc.md

```diff
@@ -1,216 +1,137 @@
+`404 Not Found`: Can indicate a missing resource. This can also indicate that the server does not support atproto, or does not support this endpoint. See response error message (or lack thereof) to clairfy.
 
-#### `500` Internal server error
+`413 Payload Too Large`: Request body was too large. If possible, split in to multiple smaller requests.
 
-The server reached an unexpected condition during processing.
+`429 Too Many Requests`: A resource limit has been exceeded, client should back off. There may be a `Retry-After` header indicating a specific back-off time period.
 
-#### `501` Method not implemented
+`500 Internal Server Error`: Generic internal service error. Client may retry after a delay.
 
-The server does not implement the requested method.
+`501 Not Implemented`: The specified endpoint is known, but not implemented. Client should *not* retry. In particular, returned when WebSockets are requested by not implemented by server.
 
-#### `502` A request to upstream failed
+`502 Bad Gateway`, `503 Service Unavailable`, or `504 Gateway Timeout`: These all usually indicate temporary or permanent service downtime. Clients may retry after a delay.
 
-The execution of the procedure depends on a call to another server which has failed.
+## Usage and Implementation Guidelines
 
-#### `503` Not enough resources
+Clients are encouraged to implement timeouts, limited retries, and randomized exponential backoff. This increases robustness in the inevitable case of sporadic downtime, while minimizing load on struggling servers.
 
-The server is under heavy load and can't complete the request.
+Servers *should* implement custom JSON error responses for all requests with an `/xrpc/` path prefix, but realistically, many services will return generic load-balancer or reverse-proxy HTML error pages. Clients should be robust to non-JSON error responses.
 
-#### `504` A request to upstream timed out
+HTTP servers and client libraries usually limit the overall size of URLs, including query parameters, and these limits constrain the use of parameters in XRPC.
 
-The execution of the procedure depends on a call to another server which timed out.
+PDS implementations are free to restrict blob uploads as they see fit. For example, they may have a global maximum size or restricted set of allowed MIME types. These should be a superset of blob constraints for all supported Lexicons.
 
-#### Remaining codes
+## Security and Privacy Considerations
 
-Any response code not explicitly enumerated should be handled as follows:
+Only HTTPS should be used over the open internet.
 
-- `1xx` treat as a `404`
-- `2xx` treat as a `200`
-- `3xx` treat as a `404` (redirects are not supported)
-- `4xx` treat as a `400`
-- `5xx` treat as a `500`
+Care should be taken with personally identifiable information in blobs, such as EXIF metadata. It is currently the *client's* responsibility to strip any sensitive EXIF metadata from blobs before uploading.
```

Comment:
Maybe add a note that a PDS _may_ strip EXIF metadata?

Seems nice for a PDS to look out for their users. I wouldn't mind putting a flag on `uploadBlob` that was something like `dontTouch` ie "don't modify these bytes at all"